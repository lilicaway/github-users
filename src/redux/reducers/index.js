import { combineReducers } from 'redux';
import usersReducer, * as gitHubUsersApi from './gitHubUsers';
import currentUserReducer, * as currentUserApi from './gitHubCurrentUserDetails';

const reducers = combineReducers({
  users: usersReducer,
  currentUser: currentUserReducer,
});

export default reducers;

// Selectors

export const getUsers = state => {
  return gitHubUsersApi.getUsers(state.users);
};

export const getUsersNextLink = state => {
  return gitHubUsersApi.getUsersNextLink(state.users);
};

export const getUsersLoadingState = state => {
  return gitHubUsersApi.getLoadingState(state.users);
};

export const getUsersLoadingErrorMessage = state => {
  return gitHubUsersApi.getErrorMessage(state.users);
};

export const getCurrentUser = state => {
  return currentUserApi.getCurrentUser(state.currentUser);
};

export const getCurrentUserLoadingState = state => {
  return currentUserApi.getCurrentUserLoadingState(state.currentUser);
};

export const getCurrentUserLoadingErrorMessage = state => {
  return currentUserApi.getCurrentUserLoadingErrorMessage(state.currentUser);
};
