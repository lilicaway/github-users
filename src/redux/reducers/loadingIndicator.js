import * as actionType from '../actions';

export const LoadingState = Object.freeze({
  INITIAL: 'INITIAL',
  LOADING: 'LOADING',
  COMPLETED: 'COMPLETED',
  ERROR: 'ERROR',
});

const createLoadingIndicatorReducer = actionPrefix => {
  return (
    state = { loadingState: LoadingState.INITIAL, errorMessage: '' },
    action
  ) => {
    switch (action.type) {
      case `${actionPrefix}_${actionType.DATA_LOADER_LOADING}`:
        return {
          ...state,
          loadingState: LoadingState.LOADING,
          errorMessage: '',
        };
      case `${actionPrefix}_${actionType.DATA_LOADER_COMPLETED}`:
        return {
          ...state,
          loadingState: LoadingState.COMPLETED,
          errorMessage: '',
        };
      case `${actionPrefix}_${actionType.DATA_LOADER_ERROR}`:
        return {
          ...state,
          loadingState: LoadingState.ERROR,
          errorMessage: action.payload,
        };
      default:
        return state;
    }
  };
};

export default createLoadingIndicatorReducer;

// Selectors

export const getLoadingState = state => {
  return state.loadingState;
};

export const getLoadingDataErrorMessage = state => {
  return state.errorMessage;
};
