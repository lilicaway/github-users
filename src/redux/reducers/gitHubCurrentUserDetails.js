import * as actionType from '../actions';
import { combineReducers } from 'redux';
import createLoadingIndicatorReducer, * as loadingIndicatorApi from './loadingIndicator';

const userReducer = (state = {}, action) => {
  switch (action.type) {
    case actionType.SET_CURRENT_USER:
      return action.payload;
    default:
      return state;
  }
};

const currentUserReducer = combineReducers({
  user: userReducer,
  loadingIndicator: createLoadingIndicatorReducer('CURRENT_USER')
});

export default currentUserReducer;

// Selectors

export class CurrentUserApi {
  constructor(state) {
    this.state = state;
  }

  getCurrentUser() {
    return this.state.user;
  }

  getLoadingState() {
    return loadingIndicatorApi.getLoadingState(this.state.loadingIndicator);
  }

  getErrorMessage() {
    return loadingIndicatorApi.getLoadingDataErrorMessage(
      this.state.loadingIndicator
    );
  }
}
