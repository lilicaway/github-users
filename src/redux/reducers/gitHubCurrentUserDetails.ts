import { combineReducers, Reducer } from 'redux';
import { GitHubUser } from '../../types/GitHubUser';
import {
  ActionType,
  DataLoaderActionSubtype,
  GitHubUsersAction
} from '../actions';
import createLoadingIndicatorReducer, * as loadingIndicatorApi from './loadingIndicator';

export type UserState = GitHubUser | null;

const userReducer: Reducer<UserState, GitHubUsersAction> = (
  state = null,
  action
) => {
  switch (action.type) {
    case ActionType.SET_CURRENT_USER:
      return action.payload;
    default:
      return state;
  }
};

interface CurrentUserState {
  user: UserState;
  loadingIndicator: loadingIndicatorApi.LoadingIndicatorState;
}

const currentUserReducer: Reducer<
  CurrentUserState,
  GitHubUsersAction
> = combineReducers({
  user: userReducer,
  loadingIndicator: createLoadingIndicatorReducer(
    DataLoaderActionSubtype.CURRENT_USER
  )
});

export default currentUserReducer;

// Selectors

export class CurrentUserApi {
  constructor(private readonly state: CurrentUserState) {}

  public getCurrentUser(): GitHubUser | undefined {
    return this.state.user || undefined;
  }

  public getLoadingState(): loadingIndicatorApi.LoadingState {
    return loadingIndicatorApi.getLoadingState(this.state.loadingIndicator);
  }

  public getErrorMessage(): string {
    return loadingIndicatorApi.getLoadingDataErrorMessage(
      this.state.loadingIndicator
    );
  }
}
