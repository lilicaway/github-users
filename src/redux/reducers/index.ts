import { combineReducers, Reducer } from 'redux';
import { GitHubUsersAction } from '../actions';
import currentUserReducer, {
  CurrentUserApi,
  CurrentUserState
} from './gitHubCurrentUserDetails';
import usersReducer, { GitHubUsersApi, UsersState } from './gitHubUsers';

export interface AppState {
  users: UsersState;
  currentUser: CurrentUserState;
}

const reducers: Reducer<AppState, GitHubUsersAction> = combineReducers({
  users: usersReducer,
  currentUser: currentUserReducer
});

export default reducers;

// Selectors

export const gitHubUsersApi = (state: AppState) => {
  return new GitHubUsersApi(state.users);
};

export const currentUserApi = (state: AppState) => {
  return new CurrentUserApi(state.currentUser);
};
