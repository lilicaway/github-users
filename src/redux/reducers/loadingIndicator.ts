import { Reducer } from 'redux';
import {
  ActionType,
  DataLoaderActionSubtype,
  GitHubUsersAction
} from '../actions';

export enum LoadingState {
  INITIAL = 'INITIAL',
  LOADING = 'LOADING',
  COMPLETED = 'COMPLETED',
  ERROR = 'ERROR'
}

export interface LoadingIndicatorState {
  loadingState: LoadingState;
  errorMessage: string;
}

const createLoadingIndicatorReducer = (
  actionSubtype: DataLoaderActionSubtype
): Reducer<LoadingIndicatorState, GitHubUsersAction> => {
  return (
    state = {
      loadingState: LoadingState.INITIAL,
      errorMessage: ''
    },
    action
  ) => {
    switch (action.type) {
      case ActionType.DATA_LOADER_LOADING:
        if (action.subType === actionSubtype) {
          return {
            ...state,
            loadingState: LoadingState.LOADING,
            errorMessage: ''
          };
        }
        break;
      case ActionType.DATA_LOADER_COMPLETED:
        if (action.subType === actionSubtype) {
          return {
            ...state,
            loadingState: LoadingState.COMPLETED,
            errorMessage: ''
          };
        }
        break;
      case ActionType.DATA_LOADER_ERROR:
        if (action.subType === actionSubtype) {
          return {
            ...state,
            loadingState: LoadingState.ERROR,
            errorMessage: action.payload
          };
        }
        break;
    }
    return state;
  };
};

export default createLoadingIndicatorReducer;

// Selectors

export const getLoadingState = (state: LoadingIndicatorState): LoadingState => {
  return state.loadingState;
};

export const getLoadingDataErrorMessage = (
  state: LoadingIndicatorState
): string => {
  return state.errorMessage;
};
