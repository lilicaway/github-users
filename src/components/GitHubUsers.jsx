import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { Image, Button, Grid, Row, Panel, Col } from 'react-bootstrap';
import LoadingIndicator from './LoadingIndicator';
import { LoadingState } from '../redux/reducers/loadingIndicator';
import { loadUsers } from '../redux/actionCreators';
import { gitHubUsersApi } from '../redux/reducers';

class GitHubUsers extends Component {
  static propTypes = {
    loadUsers: PropTypes.func.isRequired,
    loadMoreUsers: PropTypes.func.isRequired,
    users: PropTypes.array.isRequired,
    loadingState: PropTypes.string.isRequired,
    errorMessage: PropTypes.string.isRequired,
    nextLink: PropTypes.string.isRequired,
  };

  componentDidMount() {
    if (this.props.loadingState !== LoadingState.COMPLETED) {
      this.props.loadUsers();
    }
  }

  loadMoreUsers = () => {
    this.props.loadMoreUsers(this.props.nextLink);
  };

  render() {
    const userCards = this.props.users.map(user => {
      return (
        <UserCard
          key={user.id}
          username={user.login}
          avatarUrl={user.avatar_url}
        />
      );
    });
    return (
      <div>
        <LoadingIndicator
          loadingState={this.props.loadingState}
          errorMessage={this.props.errorMessage}
        />
        <Grid>
          <Row>{userCards}</Row>
          <Row>
            <Col xs={12} md={12} className="text-center">
              <Button onClick={this.loadMoreUsers}>Load More...</Button>
            </Col>
          </Row>
        </Grid>
      </div>
    );
  }
}

const mapStateToProps = (state, ownProps) => {
  const userDataApi = gitHubUsersApi(state);
  return {
    users: userDataApi.getUsers(),
    loadingState: userDataApi.getLoadingState(),
    errorMessage: userDataApi.getErrorMessage(),
    nextLink: userDataApi.getNextLink(),
  };
};

const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    loadUsers: () => {
      dispatch(loadUsers());
    },
    loadMoreUsers: nextLink => {
      dispatch(loadUsers(nextLink));
    },
  };
};

export const UnwrappedGitHubUsersForTest = GitHubUsers;
export default connect(mapStateToProps, mapDispatchToProps)(GitHubUsers);

class UserCard extends Component {
  static propTypes = {
    avatarUrl: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
  };

  render() {
    const { avatarUrl, username } = this.props;
    return (
      <Col xs={12} sm={6} md={4} lg={3}>
        <Panel className="container-fluid">
          <Image src={avatarUrl} width="100%" />
          <h3>{username}</h3>
          <p>
            <Link to={'/user/' + username}>
              <Button bsStyle="primary">User details</Button>
            </Link>
          </p>
        </Panel>
      </Col>
    );
  }
}

export const UserCardForTest = UserCard;
