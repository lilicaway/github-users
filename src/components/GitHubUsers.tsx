// import PropTypes from 'prop-types';
import * as React from 'react';
import { Button, Col, Grid, Image, Panel, Row } from 'react-bootstrap';
import {
  connect,
  MapDispatchToPropsParam,
  MapStateToPropsParam
} from 'react-redux';
import { Link } from 'react-router-dom';
import { Action } from 'redux';
import { loadUsers } from '../redux/actionCreators';
import { AppState, gitHubUsersApi } from '../redux/reducers';
import { LoadingState } from '../redux/reducers/loadingIndicator';
import { GitHubUser } from '../types/GitHubUser';
import LoadingIndicator from './LoadingIndicator';

interface Props {
  loadUsers: () => void;
  loadMoreUsers: (nextLink: string) => void;
  users: GitHubUser[];
  loadingState: string;
  errorMessage: string;
  nextLink: string;
}

class GitHubUsers extends React.Component<Props> {
  public componentDidMount() {
    if (this.props.loadingState !== LoadingState.COMPLETED) {
      this.props.loadUsers();
    }
  }

  public loadMoreUsers = () => {
    this.props.loadMoreUsers(this.props.nextLink);
  };

  public render() {
    const userCards = this.props.users.map(user => {
      return (
        <UserCard
          key={user.id}
          username={user.login}
          avatarUrl={user.avatar_url}
        />
      );
    });
    return (
      <div>
        <LoadingIndicator
          loadingState={this.props.loadingState}
          errorMessage={this.props.errorMessage}
        />
        <Grid>
          <Row>{userCards}</Row>
          <Row>
            <Col xs={12} md={12} className="text-center">
              <Button onClick={this.loadMoreUsers}>Load More...</Button>
            </Col>
          </Row>
        </Grid>
      </div>
    );
  }
}

const mapStateToProps: MapStateToPropsParam<Partial<Props>, Props, AppState> = (
  state,
  ownProps
): Partial<Props> => {
  const userDataApi = gitHubUsersApi(state);
  return {
    users: userDataApi.getUsers(),
    loadingState: userDataApi.getLoadingState(),
    errorMessage: userDataApi.getErrorMessage(),
    nextLink: userDataApi.getNextLink()
  };
};

const mapDispatchToProps: MapDispatchToPropsParam<Partial<Props>, Props> = (
  dispatch,
  ownProps
): Partial<Props> => {
  return {
    loadUsers: () => {
      // TODO: figure out a better way other than this ugly double cast.
      dispatch((loadUsers() as {}) as Action<any>);
    },
    loadMoreUsers: (nextLink: string) => {
      // TODO: figure out a better way other than this ugly double cast.
      dispatch((loadUsers(nextLink) as {}) as Action<any>);
    }
  };
};

export const UnwrappedGitHubUsersForTest = GitHubUsers;
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GitHubUsers);

interface UserCardProps {
  avatarUrl: string;
  username: string;
}

const UserCard: React.SFC<UserCardProps> = props => (
  <Col xs={12} sm={6} md={4} lg={3}>
    <Panel className="container-fluid">
      <Image src={props.avatarUrl} width="100%" />
      <h3>{props.username}</h3>
      <p>
        <Link to={'/user/' + props.username}>
          <Button bsStyle="primary">User details</Button>
        </Link>
      </p>
    </Panel>
  </Col>
);

export const UserCardForTest = UserCard;
