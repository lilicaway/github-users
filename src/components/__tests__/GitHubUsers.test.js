import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import GitHubUsers, {
  UnwrappedGitHubUsersForTest,
  UserCardForTest,
} from '../GitHubUsers';
import { Button } from 'react-bootstrap';
import { LoadingState } from '../../redux/reducers/loadingIndicator';

Enzyme.configure({ adapter: new Adapter() });

const USERS = [
  {
    login: 'mojombo',
    id: 1,
    avatar_url: 'https://avatars0.githubusercontent.com/u/1?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/mojombo',
    html_url: 'https://github.com/mojombo',
  },
  {
    login: 'defunkt',
    id: 2,
    avatar_url: 'https://avatars0.githubusercontent.com/u/2?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/defunkt',
    html_url: 'https://github.com/defunkt',
  },
  {
    login: 'pjhyett',
    id: 3,
    avatar_url: 'https://avatars0.githubusercontent.com/u/3?v=4',
    gravatar_id: '',
    url: 'https://api.github.com/users/pjhyett',
    html_url: 'https://github.com/pjhyett',
  },
];

let component;
let loadMoreUsersFunctionMock;

beforeEach(() => {
  loadMoreUsersFunctionMock = jest.fn();
  component = shallow(
    <UnwrappedGitHubUsersForTest
      users={USERS}
      loadingState={LoadingState.COMPLETED}
      errorMessage={''}
      nextLink={'someNextLink'}
      loadUsers={() => {}}
      loadMoreUsers={loadMoreUsersFunctionMock}
    />
  );
});

test('GitHub Users are shown correctly', () => {
  expect(component).toMatchSnapshot();
  expect(component.find(UserCardForTest).length).toEqual(USERS.length);
});

test('More Users can be loaded', () => {
  expect(loadMoreUsersFunctionMock.mock.calls.length).toBe(0);
  component.find(Button).simulate('click');
  expect(loadMoreUsersFunctionMock.mock.calls.length).toBe(1);
  expect(loadMoreUsersFunctionMock.mock.calls[0][0]).toBe('someNextLink');
});
