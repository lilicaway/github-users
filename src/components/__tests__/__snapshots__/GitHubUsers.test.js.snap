// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitHub Users are shown correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GitHubUsers
    errorMessage=""
    loadMoreUsers={[Function]}
    loadUsers={[Function]}
    loadingState="COMPLETED"
    nextLink="someNextLink"
    users={
        Array [
            Object {
              "avatar_url": "https://avatars0.githubusercontent.com/u/1?v=4",
              "gravatar_id": "",
              "html_url": "https://github.com/mojombo",
              "id": 1,
              "login": "mojombo",
              "url": "https://api.github.com/users/mojombo",
            },
            Object {
              "avatar_url": "https://avatars0.githubusercontent.com/u/2?v=4",
              "gravatar_id": "",
              "html_url": "https://github.com/defunkt",
              "id": 2,
              "login": "defunkt",
              "url": "https://api.github.com/users/defunkt",
            },
            Object {
              "avatar_url": "https://avatars0.githubusercontent.com/u/3?v=4",
              "gravatar_id": "",
              "html_url": "https://github.com/pjhyett",
              "id": 3,
              "login": "pjhyett",
              "url": "https://api.github.com/users/pjhyett",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <LoadingIndicator
          errorMessage=""
          loadingState="COMPLETED"
/>,
        <Grid
          bsClass="container"
          componentClass="div"
          fluid={false}
>
          <Row
                    bsClass="row"
                    componentClass="div"
          >
                    <UserCard
                              avatarUrl="https://avatars0.githubusercontent.com/u/1?v=4"
                              username="mojombo"
                    />
                    <UserCard
                              avatarUrl="https://avatars0.githubusercontent.com/u/2?v=4"
                              username="defunkt"
                    />
                    <UserCard
                              avatarUrl="https://avatars0.githubusercontent.com/u/3?v=4"
                              username="pjhyett"
                    />
          </Row>
          <Row
                    bsClass="row"
                    componentClass="div"
          >
                    <Col
                              bsClass="col"
                              className="text-center"
                              componentClass="div"
                              md={12}
                              xs={12}
                    >
                              <Button
                                        active={false}
                                        block={false}
                                        bsClass="btn"
                                        bsStyle="default"
                                        disabled={false}
                                        onClick={[Function]}
                              >
                                        Load More...
                              </Button>
                    </Col>
          </Row>
</Grid>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "errorMessage": "",
          "loadingState": "COMPLETED",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "bsClass": "container",
          "children": Array [
            <Row
              bsClass="row"
              componentClass="div"
>
              <UserCard
                            avatarUrl="https://avatars0.githubusercontent.com/u/1?v=4"
                            username="mojombo"
              />
              <UserCard
                            avatarUrl="https://avatars0.githubusercontent.com/u/2?v=4"
                            username="defunkt"
              />
              <UserCard
                            avatarUrl="https://avatars0.githubusercontent.com/u/3?v=4"
                            username="pjhyett"
              />
</Row>,
            <Row
              bsClass="row"
              componentClass="div"
>
              <Col
                            bsClass="col"
                            className="text-center"
                            componentClass="div"
                            md={12}
                            xs={12}
              >
                            <Button
                                          active={false}
                                          block={false}
                                          bsClass="btn"
                                          bsStyle="default"
                                          disabled={false}
                                          onClick={[Function]}
                            >
                                          Load More...
                            </Button>
              </Col>
</Row>,
          ],
          "componentClass": "div",
          "fluid": false,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "row",
              "children": Array [
                <UserCard
                  avatarUrl="https://avatars0.githubusercontent.com/u/1?v=4"
                  username="mojombo"
/>,
                <UserCard
                  avatarUrl="https://avatars0.githubusercontent.com/u/2?v=4"
                  username="defunkt"
/>,
                <UserCard
                  avatarUrl="https://avatars0.githubusercontent.com/u/3?v=4"
                  username="pjhyett"
/>,
              ],
              "componentClass": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "1",
                "nodeType": "function",
                "props": Object {
                  "avatarUrl": "https://avatars0.githubusercontent.com/u/1?v=4",
                  "username": "mojombo",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "2",
                "nodeType": "function",
                "props": Object {
                  "avatarUrl": "https://avatars0.githubusercontent.com/u/2?v=4",
                  "username": "defunkt",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": "3",
                "nodeType": "function",
                "props": Object {
                  "avatarUrl": "https://avatars0.githubusercontent.com/u/3?v=4",
                  "username": "pjhyett",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "bsClass": "row",
              "children": <Col
                bsClass="col"
                className="text-center"
                componentClass="div"
                md={12}
                xs={12}
>
                <Button
                                active={false}
                                block={false}
                                bsClass="btn"
                                bsStyle="default"
                                disabled={false}
                                onClick={[Function]}
                >
                                Load More...
                </Button>
</Col>,
              "componentClass": "div",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "col",
                "children": <Button
                  active={false}
                  block={false}
                  bsClass="btn"
                  bsStyle="default"
                  disabled={false}
                  onClick={[Function]}
>
                  Load More...
</Button>,
                "className": "text-center",
                "componentClass": "div",
                "md": 12,
                "xs": 12,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "active": false,
                  "block": false,
                  "bsClass": "btn",
                  "bsStyle": "default",
                  "children": "Load More...",
                  "disabled": false,
                  "onClick": [Function],
                },
                "ref": null,
                "rendered": "Load More...",
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <LoadingIndicator
            errorMessage=""
            loadingState="COMPLETED"
/>,
          <Grid
            bsClass="container"
            componentClass="div"
            fluid={false}
>
            <Row
                        bsClass="row"
                        componentClass="div"
            >
                        <UserCard
                                    avatarUrl="https://avatars0.githubusercontent.com/u/1?v=4"
                                    username="mojombo"
                        />
                        <UserCard
                                    avatarUrl="https://avatars0.githubusercontent.com/u/2?v=4"
                                    username="defunkt"
                        />
                        <UserCard
                                    avatarUrl="https://avatars0.githubusercontent.com/u/3?v=4"
                                    username="pjhyett"
                        />
            </Row>
            <Row
                        bsClass="row"
                        componentClass="div"
            >
                        <Col
                                    bsClass="col"
                                    className="text-center"
                                    componentClass="div"
                                    md={12}
                                    xs={12}
                        >
                                    <Button
                                                active={false}
                                                block={false}
                                                bsClass="btn"
                                                bsStyle="default"
                                                disabled={false}
                                                onClick={[Function]}
                                    >
                                                Load More...
                                    </Button>
                        </Col>
            </Row>
</Grid>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "errorMessage": "",
            "loadingState": "COMPLETED",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "bsClass": "container",
            "children": Array [
              <Row
                bsClass="row"
                componentClass="div"
>
                <UserCard
                                avatarUrl="https://avatars0.githubusercontent.com/u/1?v=4"
                                username="mojombo"
                />
                <UserCard
                                avatarUrl="https://avatars0.githubusercontent.com/u/2?v=4"
                                username="defunkt"
                />
                <UserCard
                                avatarUrl="https://avatars0.githubusercontent.com/u/3?v=4"
                                username="pjhyett"
                />
</Row>,
              <Row
                bsClass="row"
                componentClass="div"
>
                <Col
                                bsClass="col"
                                className="text-center"
                                componentClass="div"
                                md={12}
                                xs={12}
                >
                                <Button
                                                active={false}
                                                block={false}
                                                bsClass="btn"
                                                bsStyle="default"
                                                disabled={false}
                                                onClick={[Function]}
                                >
                                                Load More...
                                </Button>
                </Col>
</Row>,
            ],
            "componentClass": "div",
            "fluid": false,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "row",
                "children": Array [
                  <UserCard
                    avatarUrl="https://avatars0.githubusercontent.com/u/1?v=4"
                    username="mojombo"
/>,
                  <UserCard
                    avatarUrl="https://avatars0.githubusercontent.com/u/2?v=4"
                    username="defunkt"
/>,
                  <UserCard
                    avatarUrl="https://avatars0.githubusercontent.com/u/3?v=4"
                    username="pjhyett"
/>,
                ],
                "componentClass": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "1",
                  "nodeType": "function",
                  "props": Object {
                    "avatarUrl": "https://avatars0.githubusercontent.com/u/1?v=4",
                    "username": "mojombo",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "2",
                  "nodeType": "function",
                  "props": Object {
                    "avatarUrl": "https://avatars0.githubusercontent.com/u/2?v=4",
                    "username": "defunkt",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": "3",
                  "nodeType": "function",
                  "props": Object {
                    "avatarUrl": "https://avatars0.githubusercontent.com/u/3?v=4",
                    "username": "pjhyett",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "bsClass": "row",
                "children": <Col
                  bsClass="col"
                  className="text-center"
                  componentClass="div"
                  md={12}
                  xs={12}
>
                  <Button
                                    active={false}
                                    block={false}
                                    bsClass="btn"
                                    bsStyle="default"
                                    disabled={false}
                                    onClick={[Function]}
                  >
                                    Load More...
                  </Button>
</Col>,
                "componentClass": "div",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "bsClass": "col",
                  "children": <Button
                    active={false}
                    block={false}
                    bsClass="btn"
                    bsStyle="default"
                    disabled={false}
                    onClick={[Function]}
>
                    Load More...
</Button>,
                  "className": "text-center",
                  "componentClass": "div",
                  "md": 12,
                  "xs": 12,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "active": false,
                    "block": false,
                    "bsClass": "btn",
                    "bsStyle": "default",
                    "children": "Load More...",
                    "disabled": false,
                    "onClick": [Function],
                  },
                  "ref": null,
                  "rendered": "Load More...",
                  "type": [Function],
                },
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
